### PR TITLE
Add new feature to fixup accidental assignments to the same variable.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/28290
Fixes https://github.com/dotnet/roslyn/issues/31615

Looks like this:

image

Works for bogus comparisons as well:

image

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide

**Comment by mavasani** [comment]
We recently had a community contribution for a similar self assignment rule, but just for properties: see https://github.com/dotnet/roslyn-analyzers/blob/master/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidPropertySelfAssignment.cs and https://github.com/dotnet/roslyn-analyzers/blob/master/src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidPropertySelfAssignmentTests.cs.

I'll review this PR a bit later today. Thanks!

**Comment by CyrusNajmabadi** [comment]
@mavasani 

> I'll review this PR a bit later today. Thanks!

:)

**Comment by mavasani** [comment]
Ah, sorry just went off my radar. I will review it tonight or early tomorrow.

File: src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractFixAssignmentToSameVariableCodeFixProvider.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.RemoveUnusedParametersAndValues
+{
+    internal abstract class AbstractFixAssignmentToSameVariableCodeFixProvider :
```
Comment (mavasani) [inline src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractFixAssignmentToSameVariableCodeFixProvider.cs]: Redundant file that can be removed?

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractFixAssignmentToSameVariableCodeFixProvider.cs]: well that's embarrassing :)

File: src/EditorFeatures/CSharpTest/DisambiguateSameVariable/DisambiguateSameVariableTests.cs [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+    {
+        [|a = a|];
+    }
+}",
+@"class C
+{
+    int _a;
+    void M(int a)
+    {
+        _a = a;
```
Comment (RikkiGibson) [inline src/EditorFeatures/CSharpTest/DisambiguateSameVariable/DisambiguateSameVariableTests.cs]: interesting!

**Pull request merged**